Taskfile.yml

```diff
@@ -105,17 +105,46 @@ tasks:
       - mv {{.BINARY}} {{.LOCATION}}/{{.BINARY}}
 
   debug-start:
-    desc: Start headless debug
-    aliases: ["debug"]
+    desc: First, start headless debug
+    aliases: ["dbg-start", "dbg1", "debug1"]
     vars:
-      BINARY: terraform-provider-opslevel_{{.VERSION_DYNAMIC}}
+      BINARY: terraform-provider-opslevel-debug_{{.VERSION_DYNAMIC}}
+    cmds:
+      - defer: rm {{.BINARY}}
+      - go build -gcflags="all=-N -l" -ldflags="-X main.version={{.VERSION_DYNAMIC}}" -o {{.BINARY}} || exit 1
+      - chmod +x {{.BINARY}}
+      - task: terraform-init-from
+        vars: { INIT_DIR: "{{.WORKSPACE_DIR}}" }
+      - dlv exec --accept-multiclient --api-version=2 --continue --headless --wd={{.WORKSPACE_DIR}} {{.BINARY}} -- -debug | tee {{.DEBUG_LOG}}
+
+  debug-attach:
+    desc: Second, attach to headless debug sessions
+    aliases: ["dbg-attach", "dbg2", "debug2"]
+    vars:
+      LOCAL_PORT:
+        sh: grep 'API server listening at' {{.DEBUG_LOG}} | cut -d':' -f3
+    cmds:
+      - echo "Set breakpoints once connected, then run 'task debug-plan' in a new shell"
+      - dlv connect 127.0.0.1:{{.LOCAL_PORT}}
+
+  debug-apply:
+    desc: After debug-start and debug-attach, run "terraform plan" with 'TF_REATTACH_PROVIDERS'
+    aliases: ["dbg-apply"]
+    vars:
+      TF_REATTACH_PROVIDERS:
+        sh: grep 'TF_REATTACH_PROVIDERS=' {{.DEBUG_LOG}} | cut -d'=' -f2-
+    cmds:
+      - TF_REATTACH_PROVIDERS={{.TF_REATTACH_PROVIDERS}} terraform -chdir="{{.WORKSPACE_DIR}}" apply -var="OPSLEVEL_API_TOKEN=$OPSLEVEL_API_TOKEN"
+
+  debug-plan:
+    desc: After debug-start and debug-attach, run "terraform plan" with 'TF_REATTACH_PROVIDERS'
+    aliases: ["dbg-plan"]
+    vars:
+      TF_REATTACH_PROVIDERS:
+        sh: grep 'TF_REATTACH_PROVIDERS=' {{.DEBUG_LOG}} | cut -d'=' -f2-
```

Comment:
would it be possible to make this an argument to the normal `task plan` and `task apply` ?

For example on the replicated helm-chart release i do `REPLICATED_CHANNEL=Helm-Unstable task release-self-hosted`

What if we did something like `DEBUG=true task plan` or if task supports it `task plan --debug` ?

This way we wouldn't need to have separate commands.

Reply:
Maybe hacking something like this - https://taskfile.dev/usage/#forwarding-cli-arguments-to-commands

`task plan -- debug` ?

Reply:
`task plan` and `task apply` both build the local terraform provider with debug symbols removed. The setup and execution commands of these and their debug counterparts is different enough that I think it would be simpler to keep them separate

Reply:
Is there any reason we wouldn't want to always build the local provider with debug symbols.  After all - its for local testing / debugging?

Reply:
We could always build with debug symbols, but the other main issue I'm running into is working around the internal `terraform-command:` task down below.
Specifically the command in `cmds: ["terraform -chdir={{.TF_CMD_DIR}} {{.TF_COMMAND}}"]`
would need tweaking to prepend `TF_REATTACH_PROVIDERS=<stuff> terraform ... cmd` which effects most tasks

Reply:
hmm ok lets shelve this for now and maybe come back around if it gets painful to manage all 4 tasks or painful to remember there are 4